Python/Product/Analysis/DocumentBuffer.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Microsoft.PythonTools.Parsing;
+
+namespace Microsoft.PythonTools.Analysis {
+    class DocumentBuffer {
+        public int Version { get; private set; } = -1;
+        public StringBuilder Text { get; } = new StringBuilder();
+
+        public void Reset(int version, string content) {
+            Version = version;
+            Text.Clear();
+            Text.Append(content);
+        }
+
+        public void Update(IEnumerable<DocumentChangeSet> changes) {
+            foreach (var change in changes) {
+                Update(change);
+            }
+        }
+
+        public void Update(DocumentChangeSet changes) {
+            if (Version >= 0) {
+                if (changes.FromVersion < Version) {
+                    return;
+                } else if (changes.FromVersion > Version) {
+                    throw new InvalidOperationException("missing prior versions");
+                }
+            }
+
+            int lastStart = int.MaxValue;
+            var lineLoc = SplitLines(Text.ToString()).ToArray();
```

Comment:
Not sure how critical is this, but allocating new string for indexing is not required.

Reply:
I think StringBuilder is missing IndexOf[Any], and I didn't want to rewrite it. I'll mark it as a potential optimization and we can do it in its own PR.